src/mrpro/algorithms/reconstruction/IterativeSenseReconstruction.py

```diff
@@ -0,0 +1,148 @@
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import Self
+
+from mrpro.algorithms.optimizers.cg import cg
+from mrpro.algorithms.prewhiten_kspace import prewhiten_kspace
+from mrpro.algorithms.reconstruction.DirectReconstruction import DirectReconstruction
+from mrpro.algorithms.reconstruction.Reconstruction import Reconstruction
+from mrpro.data._kdata.KData import KData
+from mrpro.data.CsmData import CsmData
+from mrpro.data.DcfData import DcfData
+from mrpro.data.IData import IData
+from mrpro.data.KNoise import KNoise
+from mrpro.operators.FourierOp import FourierOp
+from mrpro.operators.LinearOperator import LinearOperator
+
+
+class IterativeSenseReconstruction(Reconstruction):
+    r"""Iterative SENSE reconstruction.
+
+    This algorithm minizes the problem
+
+    .. math:: min_x \\frac{1}{2}||W^\\frac{1}{2} (Ax - y)||_2^2
+
+    by using a conjugate gradient algorithm to solve
+
+    .. math:: H x = b
+
+    with :math:`H = A^H W A` and :math:`b = A^H W y`
+
+    where :math:`A` is the acquisition model (coil sensitivity maps, Fourier operator, k-space sampling), :math:`y` is
+    the acquired k-space data and :math:`W` describes the density compensation [1]_ .
+
+    .. [1] Pruessmann K, Weiger M, Boernert P, and Boesiger P (2001), Advances in sensitivity encoding with arbitrary
+       k-space trajectories. MRI 46, 638-651. https://doi.org/10.1002/mrm.1241
+
+    """
+
+    n_max_iter: int
+    """Maximum number of CG iterations."""
+
+    def __init__(
+        self,
+        fourier_op: LinearOperator,
+        n_iterations: int,
+        csm: CsmData | None = None,
+        noise: None | KNoise = None,
+        dcf: DcfData | None = None,
+    ) -> None:
+        """Initialize DirectReconstruction.
+
+        Parameters
+        ----------
+        fourier_op
+            Instance of the FourierOperator whose adjoint is used for reconstruction.
+        csm
+            Sensitivity maps for coil combination
+        n_iterations
+            Number of CG iterations
+        noise
+            Used for prewhitening
+        dcf
+            Density compensation. If None, no dcf will be performed.
+            Also set to None, if the FourierOperator is already density compensated.
+        """
+        super().__init__()
+        self.fourier_op = fourier_op
+        self.n_iterations = n_iterations
+        # TODO: Make this buffers once DataBufferMixin is merged
+        self.csm = csm
+        self.noise = noise
+        self.dcf = dcf
+
+    @classmethod
+    def from_kdata(cls, kdata: KData, n_iterations: int, noise: KNoise | None = None) -> Self:
```

Comment:
kdata: KData, noise: KNoise | None = None, *, n_iterations: int = 10

a) would be nice if all reconstructions.from_kdaa could work with just kdata
b) n_iterations should be last parameter, similar to combine_coils in direct reconstruction
c) n_iterations (and combine_coils) should maybe be kwargs only. ( "*" in between)